x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/TransportEnrichReindexAction.java

```diff
@@ -61,7 +64,8 @@ public TransportEnrichReindexAction(
             autoCreateIndex,
             client,
             transportService,
-            new ReindexSslConfig(settings, environment, watcherService)
+            new ReindexSslConfig(settings, environment, watcherService),
+            null
```

Comment:
Will this throw a `NullPointerException` when this special reindex task ends? I think we may have to pass an object here.

Reply:
Hi Henning, do you expect the NullPointerException to be thrown inside Reindexer->execute() ? Here, I did handle the case where we pass in a null ReindexMetrics object to the Reindexer class constructor. But perhaps you are thinking of some other scenario ?
Right now I pass in null ReindexMetrics object to TransportReindexAction when called from TransportEnrichReindexAction, but that also means we will not capture metrics for reindexing operations done by the EnrichPlugin. So I also wanted your opinion on whether we want that behavior or not.
The thing is that a metric can only be registered once on a node. So I think we can only instantiate one ReindexMetrics object for the node. If we want to record reindexing metrics from the EnrichPlugin as well, I believe we will have to instantiate the ReindexMetrics object in NodeConstruction [here](https://github.com/elastic/elasticsearch/blob/a63a2f7b008dde99dc2219220288f348616c483a/server/src/main/java/org/elasticsearch/node/NodeConstruction.java#L846) instead of as a component of ReindexPlugin like I do right now.

Reply:
I think it should work if you accept it as a parameter in `TransportEnrichReindexAction`. You expose it from the `ReindexPlugin` as a component and those are automatically injected whenever needed.

I'd prefer to do that over the null check. If there is a problem, perhaps you can let me know how it fails?

Reply:
Hi Henning, if I simply try to pass ReindexMetrics as a parameter to TransportEnrichReindexAction constructor, I run into the following failure. Could it have anything to do with the order in which the ReindexPlugin and the EnrichPlugin are created ?

> Task :docs:yamlRestTest FAILED

=== Log output of node `node{:docs:yamlRestTest-0}` ===

»    ↓ errors and warnings from /Users/ankitakumar/elastic/elasticsearch/docs/build/testclusters/yamlRestTest-0/logs/es.out ↓
» [2024-08-14T16:09:26.012689Z] [BUILD] Starting Elasticsearch process
»  Enter password for the elasticsearch keystore : Aug 14, 2024 12:09:32 PM sun.util.locale.provider.LocaleProviderAdapter <clinit>
»  WARNING: COMPAT locale provider will be removed in a future release
» [2024-08-14T12:10:03,451][ERROR][o.e.b.Elasticsearch      ] [node-0] fatal exception while booting Elasticsearch org.elasticsearch.injection.guice.CreationException: Guice creation errors:
»  
»  1) Could not find a suitable constructor in org.elasticsearch.reindex.ReindexMetrics. Classes must have either one (and only one) constructor annotated with @org.elasticsearch.injection.guice.Inject or a zero-argument constructor that is not private.
»    at org.elasticsearch.reindex.ReindexMetrics.class(Unknown Source)
»    while locating org.elasticsearch.reindex.ReindexMetrics
»      for parameter 11 at org.elasticsearch.xpack.enrich.action.TransportEnrichReindexAction.<init>(Unknown Source)
»    at _unknown_
»  
»  1 error
»       at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.injection.guice.internal.Errors.throwCreationExceptionIfErrorsExist(Errors.java:312)
»       at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.injection.guice.InjectorBuilder.initializeStatically(InjectorBuilder.java:115)
»       at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.injection.guice.InjectorBuilder.build(InjectorBuilder.java:73)
»       at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.injection.guice.Guice.createInjector(Guice.java:69)
»       at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.injection.guice.ModulesBuilder.createInjector(ModulesBuilder.java:36)
»       at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.node.NodeConstruction.construct(NodeConstruction.java:1163)
»       at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.node.NodeConstruction.prepareConstruction(NodeConstruction.java:270)
»       at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.node.Node.<init>(Node.java:193)
»       at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch$2.<init>(Elasticsearch.java:239)
»       at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.initPhase3(Elasticsearch.java:239)
»       at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:74)
»  
»  ERROR: Elasticsearch did not exit normally - check the logs at /Users/ankitakumar/elastic/elasticsearch/docs/build/testclusters/yamlRestTest-0/logs/yamlRestTest.log
»  
»  ERROR: Elasticsearch died while starting up, with exit code 1
»   ↓ last 40 non error or warning messages from /Users/ankitakumar/elastic/elasticsearch/docs/build/testclusters/yamlRestTest-0/logs/es.out ↓
» [2024-08-14T12:09:43,219][INFO ][o.e.p.PluginsService     ] [node-0] loaded module [x-pack-eql]
» [2024-08-14T12:09:43,219][INFO ][o.e.p.PluginsService     ] [node-0] loaded plugin [analysis-ukrainian]
» [2024-08-14T12:09:43,219][INFO ][o.e.p.PluginsService     ] [node-0] loaded plugin [analysis-kuromoji]
» [2024-08-14T12:09:43,219][INFO ][o.e.p.PluginsService     ] [node-0] loaded plugin [discovery-gce]
» [2024-08-14T12:09:43,219][INFO ][o.e.p.PluginsService     ] [node-0] loaded plugin [mapper-murmur3]
» [2024-08-14T12:09:43,219][INFO ][o.e.p.PluginsService     ] [node-0] loaded plugin [analysis-phonetic]
» [2024-08-14T12:09:43,219][INFO ][o.e.p.PluginsService     ] [node-0] loaded plugin [analysis-icu]
» [2024-08-14T12:09:43,219][INFO ][o.e.p.PluginsService     ] [node-0] loaded plugin [discovery-ec2]
» [2024-08-14T12:09:43,219][INFO ][o.e.p.PluginsService     ] [node-0] loaded plugin [mapper-annotated-text]
» [2024-08-14T12:09:43,219][INFO ][o.e.p.PluginsService     ] [node-0] loaded plugin [analysis-smartcn]
» [2024-08-14T12:09:43,219][INFO ][o.e.p.PluginsService     ] [node-0] loaded plugin [mapper-size]
» [2024-08-14T12:09:43,220][INFO ][o.e.p.PluginsService     ] [node-0] loaded plugin [discovery-azure-classic]
» [2024-08-14T12:09:43,220][INFO ][o.e.p.PluginsService     ] [node-0] loaded plugin [store-smb]
» [2024-08-14T12:09:43,220][INFO ][o.e.p.PluginsService     ] [node-0] loaded plugin [analysis-nori]
» [2024-08-14T12:09:43,220][INFO ][o.e.p.PluginsService     ] [node-0] loaded plugin [analysis-stempel]
» [2024-08-14T12:09:43,512][INFO ][o.e.c.u.FeatureFlag      ] [node-0] The current build is a snapshot, feature flag [failure_store] is enabled
» [2024-08-14T12:09:44,529][INFO ][o.e.c.u.FeatureFlag      ] [node-0] The current build is a snapshot, feature flag [ccs_telemetry] is enabled
» [2024-08-14T12:09:44,836][INFO ][o.e.e.NodeEnvironment    ] [node-0] using [1] data paths, mounts [[/System/Volumes/Data (/dev/disk3s5)]], net usable_space [851.2gb], net total_space [926.3gb], types [apfs]
» [2024-08-14T12:09:44,836][INFO ][o.e.e.NodeEnvironment    ] [node-0] heap size [512mb], compressed ordinary object pointers [true]
» [2024-08-14T12:09:44,858][INFO ][o.e.n.Node               ] [node-0] node name [node-0], node ID [41A1Dw_hTsGdR6iragiuYQ], cluster name [yamlRestTest], roles [remote_cluster_client, master, data_warm, data_content, transform, data_hot, ml, data_frozen, ingest, data_cold, data]
» [2024-08-14T12:09:48,502][INFO ][o.e.c.u.FeatureFlag      ] [node-0] The current build is a snapshot, feature flag [connector_secrets] is enabled
» [2024-08-14T12:09:49,400][INFO ][o.e.i.r.RecoverySettings ] [node-0] using rate limit [40mb] with [default=40mb, read=0b, write=0b, max=0b]

Reply:
If we don't care to record metrics from the TransportEnrichReindexAction, we could also call the TransportReindexAction constructor with a NOOP like so:

Reply:
Discussed this with Ryan here https://elastic.slack.com/archives/C8UUBNASY/p1724102081167389